### PR TITLE
Contract updated

### LINKED ISSUE #95: [VDA-SBT] IsClaimedSBT() and GetClaimedSBTList() update
- Can't check whether SBT is claimed with SBTType & UniqueID. Only the owner of SBT can check this
- Update to check anybody can check whit SBTType & UniqueID

Now, `isSBTClaimed(...)` function in the smart contract checks whether the sbt is claimed to the function caller (`msg.sender`). This is inconvenient in following cases:
- In the `verida-js/vda-sbt-client`, we create the vda-sbt-client with DID. This instance can call `isSBTClaimed()` function to check the claimed information regarding the DID. If there are multiple wallets for this DID, this instance can't check multiple wallets.
- In the Polygon scan (https://mumbai.polygonscan.com/ for test net), we can't call this `isSBTClaimed()` function manually, because we can't connect wallt for Read functions:
image

Same for `getClaimedSBTList()` function.

### LINKED ISSUE #99: [vda-sbt] Update for `claimSBT()` function
- Now, a wallet address can have multiple SBTs for same SBT type with different uniqueIDs. A wallet can have only one SBT per SBTType

### PR DESCRIPTION
**Pull request by ITStar10**
Next Step
	Chris should update the test/utils.ts for contract test

Closes #95

### CODE REVIEW

File: VDA-SBT/contracts/ISoulboundNFT.sol [hunk 1]
```diff
@@ -106,9 +106,10 @@ interface ISoulboundNFT {
     /**
      * @notice Get claimed SBT list of user
      * @dev Get the list of msg.sender
+     * @param didAddress address to get the claimed SBT list
      * @return uint[] Claimed token Id list
      */
-    function getClaimedSBTList() external view returns(uint[] memory);
```
Comment (ITStar10) [inline VDA-SBT/contracts/ISoulboundNFT.sol:111]: In the previous version, the `getClaimedSBTList()` function returns the claimed SBT list of the caller.
It uses the `msg.sender` in the smart contract. Here `msg.sender` points the Wallet(MetaMask for example) address that makes this transaction.
Now, the `getClaimedSBTList()` function get the claimer address as a parameter and returns the claimed sbt list of it.

File: VDA-SBT/contracts/SoulboundNFT.sol [hunk 2]
```diff
@@ -36,10 +36,10 @@ contract SoulboundNFT is VDAVerificationContract,
     EnumerableSet.StringSet private _sbtTypes;
 
     /**
-     * @notice Claimed SBT info by users
-     * @dev mapping of User => SBTType => UniqueId => tokenId
+     * @notice Claimed token id for SBTType & UniqueId
+     * @dev mapping SBTType => UniqueId => tokenId
      */
-    mapping(address => mapping(string => mapping(string => uint))) private _userInfo;
```
Comment (ITStar10) [inline VDA-SBT/contracts/SoulboundNFT.sol]: Updated the internal variables. Before, `_userInfo` variable keeps the SBTType & uniqueID per users. In the `claimSBT()` function, it checked whether the combination SBTType & uniqueID was unique to the user. This means, any other user can claim the SBT with the same SBTType & uniqueID. So removed this variable and added `_sbtID` variable. Now the contract checks whether the combination of SBTType & uniqueID is uniques in the contract.

File: VDA-SBT/test/index.ts [hunk 3]
```diff
@@ -347,12 +347,12 @@ describe("Verida Soulbound", () => {
 
     describe("Get Claimed SBT list", () => {
         it("Should return tokenId list for the claimer", async () => {
-            const idList = await contract.connect(claimer).getClaimedSBTList()
```
Comment (ITStar10) [inline VDA-SBT/test/index.ts:350]: Before, `getClaimedSBTLIst()` function returns the list of caller. So I had to call the `contract.connect(claimer)`.
Now, no need to call the `getClaimedSBTList()` function from the claimer. So removed the `connect(...)`.

**Comment by tahpot** [comment]
@ITStar10 

>Chris should update the test/utils.ts for contract test

I need more detail here. What needs to be updated?

**Review by tahpot — APPROVED** [review]

File: VDA-SBT/contracts/SoulboundNFT.sol [hunk 4]
```diff
@@ -36,10 +36,10 @@ contract SoulboundNFT is VDAVerificationContract,
     EnumerableSet.StringSet private _sbtTypes;
 
     /**
-     * @notice Claimed SBT info by users
-     * @dev mapping of User => SBTType => UniqueId => tokenId
+     * @notice Claimed token id for SBTType & UniqueId
+     * @dev mapping SBTType => UniqueId => tokenId
```
Comment (tahpot) [inline VDA-SBT/contracts/SoulboundNFT.sol]: Update this mapping

**Review by tahpot — CHANGES_REQUESTED** [review]
@ITStar10 I don't understand what this change is or why it is needed.

File: VDA-SBT/contracts/SoulboundNFT.sol (hunk 4 above)
Comment (ITStar10) [reply VDA-SBT/contracts/SoulboundNFT.sol]: Let's assum there is SBTType : "twitter" & UniqueID : "test123".
DID A claimed for this.
Can another DID claim a sbt for this SBTType & UniequeID?

**Comment by tahpot** [comment]
@ITStar10 will update this and set for review by @tahpot

**Comment by tahpot** [comment]
>Chris should update the test/utils.ts for contract test

@ITStar10 has resolved this

**Comment by tahpot** [comment]
@ITStar10 to review if the existing data structure can be kept. That will impact https://github.com/verida/verida-js/pull/296

File: VDA-SBT/ChangeLog.md [hunk 5]
```diff
@@ -1,2 +1,9 @@
 # Version 0.0.X
```
Comment (tahpot) [inline VDA-SBT/ChangeLog.md:1]: Please copy format and name of our other CHANGELOG.md files for consistency.

See example: https://github.com/verida/verida-js/blob/main/packages/client-ts/CHANGELOG.md

Please change for all blockchain repos.

File: VDA-SBT/scripts/upgrade.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
+import hre, { ethers, upgrades } from "hardhat";
```
Comment (tahpot) [inline VDA-SBT/scripts/upgrade.ts:1]: Add this upgrade info to the changelog

File: VDA-SBT/contracts/SoulboundNFT.sol
Comment (tahpot) [inline VDA-SBT/contracts/SoulboundNFT.sol:1]: >`import "@verida/name-registry/contracts/EnumerableSet.sol"`

We shouldn't have this contract rely on `name-registry`
Perhaps create a `@verida/blockchain-common` library and place `EnumerableSet` in there, then link to it?

>`import "@verida/vda-verification-contract/contracts/VDAVerificationContract.sol";`

This needs to be in `package.json`? I can't run tests without yarn linking `@verida/vda-verification-contract`

**Review by tahpot — CHANGES_REQUESTED** [review]
The code looks good, but I can't run the tests as there is an issue compiling `VDA-Name-Registry`:

```
VDA-Name-Registry % yarn run build
yarn run v1.22.15
warning ../../package.json: No license field
$ npx hardhat compile
Error HH8: There's one or more errors in your config file:

  * Invalid account: #0 for network: bsctestnet - Expected string, received undefined
  * Invalid account: #0 for network: bscmainnet - Expected string, received undefined
  * Invalid account: #0 for network: polygonmainnet - Expected string, received undefined
  * Invalid account: #0 for network: polygontestnet - Expected string, received undefined
  * Invalid value undefined for HardhatConfig.networks.polygonprivatenet.url - Expected a value of type string.
  * Invalid account: #0 for network: polygonprivatenet - Expected string, received undefined
```

File: VDA-SBT/ChangeLog.md (hunk 5 above)
Comment (tahpot) [reply VDA-SBT/ChangeLog.md:1]: Issue was lack of `.env` file.

@ITStar10 Please add instructions (and an example file) on how to configure `.env` file for all blockchain packages.

File: VDA-SBT/contracts/SoulboundNFT.sol
Comment (tahpot) [reply VDA-SBT/contracts/SoulboundNFT.sol:1]: Do this after merging all the PR's

**Review by tahpot — APPROVED** [review]

**Pull request merged**